--- a/root/wifibroadcast/rssitx.c
+++ b/root/wifibroadcast/rssitx.c
@@ -213,10 +213,10 @@ void sendRSSI(int sock, telemetry_data_t
     		fclose(fp);
 		framedata.cpuload = (((b[0]+b[1]+b[2]) - (a[0]+a[1]+a[2])) / ((b[0]+b[1]+b[2]+b[3]) - (a[0]+a[1]+a[2]+a[3]))) * 100;
 
-		fp = fopen("/sys/class/thermal/thermal_zone0/temp","r");
-		int temp = 0;
-		fscanf(fp,"%d",&temp);
-		fclose(fp);
+		//fp = fopen("/sys/class/thermal/thermal_zone0/temp","r");
+		int temp = 25;
+		//fscanf(fp,"%d",&temp);
+		//fclose(fp);
 		//fprintf(stderr,"temp: %d\n",temp/1000);
 		framedata.temp = temp / 1000;
 	}
