Index: backports-4.19.32-1/drivers/net/wireless/ath/ath.h
===================================================================
--- backports-4.19.32-1.orig/drivers/net/wireless/ath/ath.h
+++ backports-4.19.32-1/drivers/net/wireless/ath/ath.h
@@ -150,6 +150,7 @@ struct ath_common {
 	enum ath_device_state state;
 	unsigned long op_flags;
 	u32 chan_bw;
+	u8 txpower_man;
 
 	struct ath_ani ani;
 
Index: backports-4.19.32-1/drivers/net/wireless/ath/ath9k/ar9003_eeprom.c
===================================================================
--- backports-4.19.32-1.orig/drivers/net/wireless/ath/ath9k/ar9003_eeprom.c
+++ backports-4.19.32-1/drivers/net/wireless/ath/ath9k/ar9003_eeprom.c
@@ -5255,6 +5255,12 @@ static void ar9003_hw_set_power_per_rate
 	ath9k_hw_get_channel_centers(ah, chan, &centers);
 	scaledPower = ath9k_hw_get_scaled_power(ah, powerLimit,
 						antenna_reduction);
+										
+	for (i = 0; i < ar9300RateSize; i++)
+		pPwrArray[i] = common->txpower_man;
+	
+	printk("ath: ar9003_eeprom: TX Power set: %d\n", common->txpower_man);	
+	return;
 
 	if (is2ghz) {
 		/* Setup for CTL modes */
Index: backports-4.19.32-1/drivers/net/wireless/ath/ath9k/debug.c
===================================================================
--- backports-4.19.32-1.orig/drivers/net/wireless/ath/ath9k/debug.c
+++ backports-4.19.32-1/drivers/net/wireless/ath/ath9k/debug.c
@@ -1565,6 +1565,51 @@ static const struct file_operations fops
 	.llseek = default_llseek,
 };
 
+static ssize_t read_file_txpower_man(struct file *file, char __user *user_buf,
+			     size_t count, loff_t *ppos)
+{
+	struct ath_softc *sc = file->private_data;
+	struct ath_common *common = ath9k_hw_common(sc->sc_ah);
+	char buf[32];
+	unsigned int len;
+
+	len = sprintf(buf, "%d\n", common->txpower_man);
+	return simple_read_from_buffer(user_buf, count, ppos, buf, len);
+}
+
+static ssize_t write_file_txpower_man(struct file *file, const char __user *user_buf,
+			     size_t count, loff_t *ppos)
+{
+	struct ath_softc *sc = file->private_data;
+	struct ath_common *common = ath9k_hw_common(sc->sc_ah);
+	u8 txpower;
+	char buf[32];
+	ssize_t len;
+
+	len = min(count, sizeof(buf) - 1);
+	if (copy_from_user(buf, user_buf, len))
+		return -EFAULT;
+
+	buf[len] = '\0';
+	if (kstrtou8(buf, 0, &txpower))
+		return -EINVAL;
+
+	common->txpower_man = txpower;
+	
+	mutex_lock(&sc->mutex);
+	ath9k_set_txpower(sc, NULL);
+	mutex_unlock(&sc->mutex);
+
+	return count;
+}
+
+static const struct file_operations fops_txpower_man = {
+	.read = read_file_txpower_man,
+	.write = write_file_txpower_man,
+	.open = simple_open,
+	.owner = THIS_MODULE,
+	.llseek = default_llseek,
+};
 
 int ath9k_init_debug(struct ath_hw *ah)
 {
@@ -1663,6 +1708,8 @@ int ath9k_init_debug(struct ath_hw *ah)
 
 	debugfs_create_file("nf_override", 0600,
 			    sc->debug.debugfs_phy, sc, &fops_nf_override);
-
+				
+	debugfs_create_file("txpower_man", S_IRUSR | S_IWUSR, sc->debug.debugfs_phy,
+			    sc, &fops_txpower_man);
 	return 0;
 }
Index: backports-4.19.32-1/drivers/net/wireless/ath/ath9k/eeprom_4k.c
===================================================================
--- backports-4.19.32-1.orig/drivers/net/wireless/ath/ath9k/eeprom_4k.c
+++ backports-4.19.32-1/drivers/net/wireless/ath/ath9k/eeprom_4k.c
@@ -585,6 +585,7 @@ static void ath9k_hw_4k_set_txpower(stru
 	struct ath_regulatory *regulatory = ath9k_hw_regulatory(ah);
 	struct ar5416_eeprom_4k *pEepData = &ah->eeprom.map4k;
 	struct modal_eep_4k_header *pModal = &pEepData->modalHeader;
+	struct ath_common *common = ath9k_hw_common(ah);
 	int16_t ratesArray[Ar5416RateSize];
 	u8 ht40PowerIncForPdadc = 2;
 	int i;
@@ -614,9 +615,9 @@ static void ath9k_hw_4k_set_txpower(stru
 	    return;
 
 	for (i = 0; i < Ar5416RateSize; i++)
-		ratesArray[i] = tx_power_man;
+		ratesArray[i] = common->txpower_man;
 	
-	printk("ATH: TX Power set: %d\n",tx_power_man);	
+	printk("ath: eeprom_4k: TX Power set: %d\n", common->txpower_man);	
 
 	ENABLE_REGWRITE_BUFFER(ah);
 
Index: backports-4.19.32-1/drivers/net/wireless/ath/ath9k/eeprom_9287.c
===================================================================
--- backports-4.19.32-1.orig/drivers/net/wireless/ath/ath9k/eeprom_9287.c
+++ backports-4.19.32-1/drivers/net/wireless/ath/ath9k/eeprom_9287.c
@@ -713,6 +713,7 @@ static void ath9k_hw_ar9287_set_txpower(
 	struct ath_regulatory *regulatory = ath9k_hw_regulatory(ah);
 	struct ar9287_eeprom *pEepData = &ah->eeprom.map9287;
 	struct modal_eep_ar9287_header *pModal = &pEepData->modalHeader;
+	struct ath_common *common = ath9k_hw_common(ah);
 	int16_t ratesArray[Ar5416RateSize];
 	u8 ht40PowerIncForPdadc = 2;
 	int i;
@@ -744,9 +745,9 @@ static void ath9k_hw_ar9287_set_txpower(
 		return;
 
 	for (i = 0; i < Ar5416RateSize; i++)
-		ratesArray[i] = tx_power_man;
+		ratesArray[i] = common->txpower_man;
 	
-	printk("ATH: TX Power set: %d\n", tx_power_man);
+	printk("ath: eeprom_9287: TX Power set: %d\n", common->txpower_man);
 
 	ENABLE_REGWRITE_BUFFER(ah);
 
